hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -557,9 +564,20 @@ SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long
                     partitionKeyComponents[i] = perMapCandidate.get(attribute).eval(null, evalContext);
                 }
 
-                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
-                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
-                );
+                // Mimic calculation performed for IMap put/get operations
+                // in AbstractSerializationService.calculatePartitionHash.
+                // IMap partitioning strategy is passed there.
+                //
+                // We cannot pass AttributePartitioningStrategy because we do not full key object, but
+                // IDENTITY_PARTITIONING_STRATEGY on partitionKeyComponents will give the same result
+                // as AttributePartitioningStrategy would on a full key.
+                // For other IMap strategies we use IMap strategy (only Default is supported)
+                // or null which will use global strategy.
+                Data keyData = serializationService.toData(
+                        // constructAttributeBasedKey gives needed result also for __key = ?
+                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents),
+                        strategy instanceof AttributePartitioningStrategy ? IDENTITY_PARTITIONING_STRATEGY : strategy);
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(keyData);
                 if (partition == null) {
                     // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
```

Comment:
actually what is the problem with cluster being mid-repartitioning/migration? We can calculate partition id regardless of the state of the cluster. Maybe to job will start on a wrong member which is not the owner of the partition, but the processors should be (?) migration aware.